src/stdlib.c

```diff
@@ -115,112 +115,113 @@ static void base64_decode(const unsigned char *s, int len, char *dst) {
-    base64_decode((const unsigned char *) v->v.str.buf, (int) v->v.str.len,
-                  result->v.str.buf);
+static val_t b64_transform(struct v7 *v7, val_t this_obj, val_t args,
+                           void (func)(const unsigned char *, int, char *),
+                           double mult) {
+  val_t arg0 = v7_array_at(v7, args, 0);
+  val_t res = v7_create_undefined();
+
+  (void) this_obj;
+  if (v7_is_string(arg0)) {
+    size_t n;
+    const char *s = v7_to_string(v7, &arg0, &n);
+    char *buf = (char *) malloc(n * mult + 2);
+    if (buf != NULL) {
+      func((const unsigned char *) s, (int) n, buf);
+      res = v7_create_string(v7, buf, strlen(buf), 1);
+      free(buf);
+    }
   }
-  return V7_OK;
+
+  return res;
 }
 
-V7_PRIVATE enum v7_err Std_base64_encode(struct v7_c_func_arg *cfa) {
-  struct v7_val *v = cfa->args[0], *result;
+static val_t Std_base64_decode(struct v7 *v7, val_t this_obj, val_t args) {
+  return b64_transform(v7, this_obj, args, base64_decode, 0.75);
+}
 
-  result = v7_push_string(cfa->v7, NULL, 0, 0);
-  if (cfa->num_args == 1 && v->type == V7_TYPE_STR && v->v.str.len > 0) {
-    result->v.str.len = v->v.str.len * 3 / 2 + 1;
-    result->v.str.buf = (char *) malloc(result->v.str.len + 1);
-    base64_encode((const unsigned char *) v->v.str.buf, (int) v->v.str.len,
-                  result->v.str.buf);
-  }
-  return V7_OK;
+static val_t Std_base64_encode(struct v7 *v7, val_t this_obj, val_t args) {
+  return b64_transform(v7, this_obj, args, base64_encode, 1.5);
 }
 
 #ifndef V7_NO_FS
-V7_PRIVATE enum v7_err Std_load(struct v7_c_func_arg *cfa) {
-  int i;
-  struct v7_val *obj = v7_push_new_object(cfa->v7);
-
-  /* Push new object as a context for the loading new module */
-  obj->next = cfa->v7->ctx;
-  cfa->v7->ctx = obj;
-
-  for (i = 0; i < cfa->num_args; i++) {
-    if (v7_type(cfa->args[i]) != V7_TYPE_STR) return V7_TYPE_ERROR;
-    if (!v7_exec_file(cfa->v7, cfa->args[i]->v.str.buf)) return V7_ERROR;
+static val_t Std_load(struct v7 *v7, val_t this_obj, val_t args) {
+  val_t arg0 = v7_array_at(v7, args, 0);
+  val_t res = v7_create_undefined();
+
+  (void)this_obj;
+  if (v7_is_string(arg0)) {
+    size_t n;
+    const char *s = v7_to_string(v7, &arg0, &n);
+    v7_exec_file(v7, &res, s);
   }
 
-  /* Pop context, and return it */
-  cfa->v7->ctx = obj->next;
-  v7_push_val(cfa->v7, obj);
-
-  return V7_OK;
+  return res;
 }
 
-V7_PRIVATE enum v7_err Std_read(struct v7_c_func_arg *cfa) {
-  struct v7_val *v;
+static val_t Std_read(struct v7 *v7, val_t this_obj, val_t args) {
+  val_t arg0 = v7_array_at(v7, args, 0);
   char buf[2048];
   size_t n;
 
-  if ((v = v7_get(cfa->this_obj, "fp")) != NULL &&
-      (n = fread(buf, 1, sizeof(buf), (FILE *)(unsigned long) v->v.num)) > 0) {
-    v7_push_string(cfa->v7, buf, n, 1);
-  } else {
-    v7_make_and_push(cfa->v7, V7_TYPE_NULL);
+  (void)this_obj;
+  if (v7_is_double(arg0)) {
+    FILE *fp = (FILE *) (uintptr_t) v7_to_double(arg0);
```

Comment:
Changed FILE *fp to int fd

On Fri, Feb 27, 2015 at 5:36 PM, Marko Mikulicic <notifications@github.com>
wrote:

> In src/stdlib.c
> <https://github.com/cesanta/v7/pull/305#discussion_r25523105>:
>
> >    char buf[2048];
> >    size_t n;
> >
> > -  if ((v = v7_get(cfa->this_obj, "fp")) != NULL &&
> > -      (n = fread(buf, 1, sizeof(buf), (FILE *)(unsigned long) v->v.num)) > 0) {
> > -    v7_push_string(cfa->v7, buf, n, 1);
> > -  } else {
> > -    v7_make_and_push(cfa->v7, V7_TYPE_NULL);
> > +  (void)this_obj;
> > +  if (v7_is_double(arg0)) {
> > +    FILE *fp = (FILE *) (uintptr_t) v7_to_double(arg0);
>
> shouldn't we use foreign pointers here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cesanta/v7/pull/305/files#r25523105>.
>